rclcpp/src/rclcpp/time_source.cpp

```diff
@@ -242,6 +250,24 @@ void TimeSource::create_clock_sub()
       rclcpp::QosPolicyKind::Reliability,
     });
 
+  if (use_clock_thread_) {
+    clock_callback_group_ = node_base_->create_callback_group(
+      rclcpp::CallbackGroupType::MutuallyExclusive,
+      false
+    );
+    options.callback_group = clock_callback_group_;
+    if (!clock_executor_thread_.joinable()) {
+      clock_executor_thread_ = std::thread(
+        [this]() {
+          cancel_clock_executor_promise_ = std::promise<void>{};
+          auto future = cancel_clock_executor_promise_.get_future();
+          clock_executor_.add_callback_group(clock_callback_group_, node_base_);
+          clock_executor_.spin_until_future_complete(future);
```

Comment:
I'm wondering if the future is necessary? You already cancel the executor below, so that should interrupt any spinning. You could just do something like this:

```c++
while (clock_not_canceled_) {
  clock_executor_.spin_once();
}

// And in the destroy_clock_sub() method...
clock_not_canceled_ = false;
clock_executor_.cancel();
clock_executor_thread_.join();
// ... and so on
```

It's not much different, and this code is certainly not wrong, but the future seems unnecessary to me.

Just a question, not a request to change it.

Reply:
The reason is that the class destructor maybe called before `executor.spin()` (in `create_clock_sub` function). In this case, since `executor.cancel()` only cancels the executor **already** spinning, the executor will be always spinning once `executor.spin()` is reached. 
This [conversation](https://github.com/ros2/rclcpp/pull/1556#discussion_r594604635) explains the issue. 
And [this other one](https://github.com/ros2/rclcpp/pull/1556#discussion_r595632418) explains why the design with`spin_until_future_complete` is better that just using a flag to get executor state.